ocaml/xapi/xapi_observer.ml

```diff
@@ -0,0 +1,106 @@
+(*
+ * Copyright (C) 2023 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+let create ~__context ~name_label ~name_description ~hosts ~attributes
+    ~endpoints ~components ~enabled =
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
```

Comment:
I think we need some input validation here for `hosts`, `endpoints`, `attributes` and `components`. And also in the setters below.

Reply:
For the URL, I think `Uri.of_string` is better than nothing. For the attributes, we can probably check for illegal characters in keys and values? You could add a function to the library to do that. We can be quite strict initially and only accept e.g. letters, numbers and some essential special characters.

Reply:
All validators done, endpoints currently will wait until a better method of validation comes about.